### PR TITLE
Static IP address for WiFi

### PR DESCRIPTION
**Pull request by gamblor21**
As mentioned in issue #6274 there was a request to add the ability to statically set your IP address. Using @anecdata's suggestions I have added in this functionality. There are three new methods for wifi.Radio:

1. `stop_dhcp` - Stop the DHCP client (required to set a static address)
2. `start_dhcp` - Start the DHCP client (will automatically grab an IP address if it was stopped)
3. `set_ipv4_address` - Provide the static IP, subnet and gateway addresses (all are required). This will stop the DHCP client automatically as it is required.

One thing to note, if you stop DHCP and do not set a static IP address and a call to connect to WiFi will fail.
The set_ipv4_address is a callable function (as opposed to a setter) as it does require all 3 values that the current getters can retrieve.

### CODE REVIEW

**Review by anecdata — COMMENTED** [review]
Thanks for diving in to this!

Using this test code on a
`Adafruit CircuitPython 8.0.0-alpha.0-22-g297aa91f2 on 2022-05-27; Adafruit QT Py ESP32S2 with ESP32S2`
to set up the static parameters:
```
import time
import wifi
import ipaddress
from secrets import secrets

print(f"Setting IPv4...")
wifi.radio.set_ipv4_address(ipaddress.ip_address("192.168.6.252"),
                            ipaddress.ip_address("192.168.4.1"),
                            ipaddress.ip_address("255.255.252.0"))
print(f"IPv4    {wifi.radio.ipv4_address}")  # None
print(f"Gateway {wifi.radio.ipv4_gateway}")  # None
print(f"Subnet  {wifi.radio.ipv4_subnet}")   # None

print(f"Connecting...")
wifi.radio.connect(secrets['ssid'], secrets['password'])
print(f"IPv4    {wifi.radio.ipv4_address}")  # 192.168.6.252
print(f"Gateway {wifi.radio.ipv4_gateway}")  # 192.168.4.1
print(f"Subnet  {wifi.radio.ipv4_subnet}")   # 255.255.252.0

print(f"DNS     {wifi.radio.ipv4_dns}")      # 0.0.0.0
print(f"RSSI    {wifi.radio.ap_info.rssi}")  # e.g., -65
```
Device is able to connect to the AP, and the static parameters are correctly reflected. TCP sockets and Requests work on LAN numeric IPv4 addresses. It's not able to `ping` LAN or WAN numeric IPv4 addresses for some reason (result always is `None`) - I may be missing a step or it could be a local network issue. Without DHCP, there is no DNS address. I don't think we have a way to set DNS address(es). So there is no name resolution. But, Requests to an mDNS host does work :-)

I haven't gotten WAN Requests (v1.11.2) to a numeric IPv4 address to work yet:
```
Traceback (most recent call last):
  File "code.py", line ..., in <module>
  File "adafruit_requests.py", line 720, in get
  File "adafruit_requests.py", line 661, in request
  File "adafruit_requests.py", line 512, in _get_socket
RuntimeError: Sending request failed
```

File: shared-bindings/wifi/Radio.c [hunk 1]
```diff
@@ -459,6 +480,26 @@ STATIC mp_obj_t wifi_radio_get_ap_info(mp_obj_t self) {
 }
 MP_DEFINE_CONST_FUN_OBJ_1(wifi_radio_get_ap_info_obj, wifi_radio_get_ap_info);
 
+//|     def wifi_radio_start_dhcp_client(self) -> None:
```
Comment (tannewt) [inline shared-bindings/wifi/Radio.c:483]:
```suggestion
//|     def start_dhcp(self) -> None:
```

File: shared-bindings/wifi/Radio.c [hunk 2]
```diff
@@ -459,6 +480,26 @@ STATIC mp_obj_t wifi_radio_get_ap_info(mp_obj_t self) {
...
+//|         """Starts the DHCP client."""
+//|         ...
+//|
+STATIC mp_obj_t wifi_radio_start_dhcp_client(mp_obj_t self) {
+    common_hal_wifi_radio_start_dhcp_client(self);
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_1(wifi_radio_start_dhcp_client_obj, wifi_radio_start_dhcp_client);
+
+//|     def wifi_radio_stop_dhcp_client(self) -> None:
```
Comment (tannewt) [inline shared-bindings/wifi/Radio.c:493]:
```suggestion
//|     def stop_dhcp(self) -> None:
```

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for adding this! I can't think of any changes but did see two minor doc issues.

**Comment by gamblor21** [comment]
> It's not able to `ping` LAN or WAN numeric IPv4 addresses for some reason (result always is `None`) - I may be missing a step or it could be a local network issue. 

Ping is working for me, so not sure what could be wrong there?

> Without DHCP, there is no DNS address. I don't think we have a way to set DNS address(es). So there is no name resolution. But, Requests to an mDNS host does still work, and the mDNS host reflects the static IP :-)

Well that won't do! I added a set method for DNS. Totally skipped my mind as I rarely use DNS for CP projects.

File: shared-bindings/wifi/Radio.c [hunk 3]
```diff
@@ -414,6 +414,27 @@ MP_DEFINE_CONST_FUN_OBJ_1(wifi_radio_get_ipv4_subnet_ap_obj, wifi_radio_get_ipv4
 MP_PROPERTY_GETTER(wifi_radio_ipv4_subnet_ap_obj,
     (mp_obj_t)&wifi_radio_get_ipv4_subnet_ap_obj);
 
+//|     def set_ipv4_address(self, ipv4: ipaddress.IPv4Address, netmask: ipaddress.IPv4Address, gateway: ipaddress.IPv4Address) -> None:
```
Comment (tannewt) [inline shared-bindings/wifi/Radio.c:417]: Want to add DNS as optional here too? Fine to leave the property as it is now too.

File: shared-bindings/wifi/Radio.c [hunk 4]
```diff
@@ -414,23 +414,24 @@ MP_DEFINE_CONST_FUN_OBJ_1(wifi_radio_get_ipv4_subnet_ap_obj, wifi_radio_get_ipv4
 MP_PROPERTY_GETTER(wifi_radio_ipv4_subnet_ap_obj,
     (mp_obj_t)&wifi_radio_get_ipv4_subnet_ap_obj);
 
-//|     def set_ipv4_address(self, ipv4: ipaddress.IPv4Address, netmask: ipaddress.IPv4Address, gateway: ipaddress.IPv4Address) -> None:
-//|         """Sets the IP v4 address of the station. Must include the netmask and gateway.
+//|     def set_ipv4_address(self, ipv4: ipaddress.IPv4Address, netmask: ipaddress.IPv4Address, gateway: ipaddress.IPv4Address, ipv4_dns: Optional[ipaddress.IPv4Address]) -> None:
```
Comment (tannewt) [inline shared-bindings/wifi/Radio.c:417]: Sorry, one more thing. Generally when a function takes multiple arguments of the same type, it's best to force them to be keyword arguments. That way they are labeled. So, here you'd do:
```suggestion
//|     def set_ipv4_address(self, *, ipv4: ipaddress.IPv4Address, netmask: ipaddress.IPv4Address, gateway: ipaddress.IPv4Address, ipv4_dns: Optional[ipaddress.IPv4Address]) -> None:
```

File: shared-bindings/wifi/Radio.c [hunk 5]
```diff
@@ -414,23 +414,24 @@ MP_DEFINE_CONST_FUN_OBJ_1(wifi_radio_get_ipv4_subnet_ap_obj, wifi_radio_get_ipv4
...
 //|            Setting the address manually will stop the DHCP client."""
 //|         ...
 STATIC mp_obj_t wifi_radio_set_ipv4_address(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
-    enum { ARG_ipv4, ARG_netmask, ARG_gateway };
+    enum { ARG_ipv4, ARG_netmask, ARG_gateway, ARG_ipv4_dns };
     static const mp_arg_t allowed_args[] = {
         { MP_QSTR_ipv4, MP_ARG_REQUIRED | MP_ARG_OBJ, },
         { MP_QSTR_netmask, MP_ARG_REQUIRED | MP_ARG_OBJ, },
         { MP_QSTR_gateway, MP_ARG_REQUIRED | MP_ARG_OBJ, },
+        { MP_QSTR_ipv4_dns, MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
```
Comment (tannewt) [inline shared-bindings/wifi/Radio.c:427]:
```suggestion
        { MP_QSTR_ipv4, MP_ARG_REQUIRED | MP_ARG_KW_ONLY | MP_ARG_OBJ, },
        { MP_QSTR_netmask, MP_ARG_REQUIRED | MP_ARG_KW_ONLY | MP_ARG_OBJ, },
        { MP_QSTR_gateway, MP_ARG_REQUIRED | MP_ARG_KW_ONLY | MP_ARG_OBJ, },
        { MP_QSTR_ipv4_dns, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
```

**Review by tannewt — CHANGES_REQUESTED** [review]
Sorry, I thought of one more thing. Thank you for your patience.

File: shared-bindings/wifi/Radio.c (hunk 5 above)
Comment (gamblor21) [reply shared-bindings/wifi/Radio.c]: No problem. Easy changes to make.

**Review by tannewt — APPROVED** [review]
Thank you!

**Pull request merged**